File: Data Structures/Stack/C/stack.c
#include <stdio.h>
#include <limits.h>
#include <stdlib.h>

struct Stack
{
    int *items;
    unsigned int capacity;
    int head;
};

// Stack is full when head is equal to the last index
int isFull(struct Stack *stack)
{
    return stack->head == stack->capacity - 1;
}

// Stack is empty when head is equal to -1
int isEmpty(struct Stack *stack)
{
    return stack->head == -1;
}

// Function to add an item to stack.  It increases head by 1
void push(struct Stack *stack, int item)
{
    if (isFull(stack))
        return;
    stack->items[++stack->head] = item;
}

// Function to remove an item from stack.  It decreases head by 1
int pop(struct Stack *stack)
{
    if (isEmpty(stack))
        return INT_MIN;
    return stack->items[stack->head--];
}

// Function to return the head from stack without removing it
int peek(struct Stack *stack)
{
    if (isEmpty(stack))
        return INT_MIN;
    return stack->items[stack->head];
}

struct Stack *makeStack(unsigned capacity)
{
    struct Stack *stack = (struct Stack *)malloc(sizeof(struct Stack));
    stack->capacity = capacity;
    stack->head = -1;
    stack->items = (int *)malloc(stack->capacity * sizeof(int));
    return stack;
}

int main()
{
    struct Stack *stack = makeStack(100);

    push(stack, 10);
    printf("%d pushed to stack\n", 10);
    push(stack, 20);
    printf("%d pushed to stack\n", 20);
    push(stack, 30);
    printf("%d pushed to stack\n", 30);

    printf("%d popped from stack\n", pop(stack));
    return 1;
}

File: Data Structures/Stack/Java/Stack.java
class Stack { 
	static final int MAX = 1000; 
	int top; 
	int a[] = new int[MAX];

	boolean isEmpty() { 
		return (top < 0); 
	} 

	Stack() { 
		top = -1; 
	} 

	boolean push(int x) { 
		if (top >= (MAX - 1)) { 
			System.out.println("Stack Overflow"); 
			return false; 
		} else { 
			a[++top] = x; 
			System.out.println(x + " pushed into stack"); 
			return true; 
		} 
	} 

	int pop() { 
		if (top < 0) { 
			System.out.println("Stack Underflow"); 
			return 0; 
		} lse { 
			int x = a[top--]; 
			return x; 
		} 
	} 

	int peek() { 
		if (top < 0) { 
			System.out.println("Stack Underflow"); 
			return 0; 
		} else { 
			int x = a[top]; 
			return x; 
		} 
	} 
} 

class Main { 
	public static void main(String args[]) { 
		Stack s = new Stack(); 
		s.push(10); 
		s.push(20); 
		s.push(30); 
		System.out.println(s.pop() + " Popped from stack"); 
	} 
} 
